Makefile

```diff
@@ -32,54 +36,76 @@ ifneq ($(DEPRECATION_ON),0)
 endif
 # CXXFLAGS += -DTEXTURE_DEBUG
 
-LDFLAGS := -lstdc++ -lm -ldl -lpng
+LDFLAGS := -lm -ldl -lpng 
+
+ifneq ($(LLD),0)
+  LDFLAGS += -fuse-ld=lld
+endif
 
 UNAME := $(shell uname)
 ifneq ($(UNAME), Darwin)
     LDFLAGS += -Wl,-export-dynamic -lstdc++fs
 endif
 
-SRC_DIRS := ZAPD ZAPD/ZRoom ZAPD/ZRoom/Commands ZAPD/Overlays ZAPD/HighLevel ZAPD/Utils
+ZAPD_SRC_DIRS := $(shell find ZAPD -type d)
+SRC_DIRS = $(ZAPD_SRC_DIRS) lib/tinyxml2
 
-ZAPD_CPP_FILES := $(foreach dir,$(SRC_DIRS),$(wildcard $(dir)/*.cpp))
-ZAPD_H_FILES   := $(foreach dir,$(SRC_DIRS),$(wildcard $(dir)/*.h))
+ZAPD_CPP_FILES := $(foreach dir,$(ZAPD_SRC_DIRS),$(wildcard $(dir)/*.cpp))
+ZAPD_H_FILES   := $(foreach dir,$(ZAPD_SRC_DIRS),$(wildcard $(dir)/*.h))
 
 CPP_FILES += $(ZAPD_CPP_FILES) lib/tinyxml2/tinyxml2.cpp
-O_FILES   := $(CPP_FILES:.cpp=.o)
+O_FILES   := $(foreach f,$(CPP_FILES:.cpp=.o),build/$f)
+O_FILES   += build/ZAPD/BuildInfo.o
+
+# create build directories
+$(shell mkdir -p $(foreach dir,$(SRC_DIRS),build/$(dir)))
+
 
+# Main targets
 all: ZAPD.out copycheck
 
-genbuildinfo:
+build/ZAPD/BuildInfo.o:
 	python3 ZAPD/genbuildinfo.py $(COPYCHECK_ARGS)
+	$(CXX) $(CXXFLAGS) $(OPTFLAGS) $(INC) -c build/ZAPD/BuildInfo.cpp -o build/ZAPD/BuildInfo.o
 
 copycheck: ZAPD.out
 	python3 copycheck.py
 
 clean:
-	rm -f $(O_FILES) ZAPD.out
+	rm -rf build ZAPD.out
 	$(MAKE) -C lib/libgfxd clean
+	$(MAKE) -C ZAPDUtils clean
+	$(MAKE) -C ExporterTest clean
 
 rebuild: clean all
 
 format:
 	clang-format-11 -i $(ZAPD_CPP_FILES) $(ZAPD_H_FILES)
+	$(MAKE) -C ZAPDUtils format
+	$(MAKE) -C ExporterTest format
 
-.PHONY: all genbuildinfo copycheck clean rebuild format
+.PHONY: all build/ZAPD/BuildInfo.o copycheck clean rebuild format
 
-%.o: %.cpp
+build/%.o: %.cpp
+	$(CXX) $(CXXFLAGS) $(OPTFLAGS) $(INC) -c $< -o $@
+
+build/ZAPD/Main.o: ZAPD/Main.cpp build/ZAPD/BuildInfo.o
 	$(CXX) $(CXXFLAGS) $(OPTFLAGS) $(INC) -c $< -o $@
 
-ZAPD/Main.o: genbuildinfo ZAPD/Main.cpp
-	$(CXX) $(CXXFLAGS) $(OPTFLAGS) $(INC) -c ZAPD/Main.cpp -o $@
 
+# Submakes
 lib/libgfxd/libgfxd.a:
 	$(MAKE) -C lib/libgfxd
 
-ExporterTest/ExporterTest.a:
+.PHONY: ExporterTest
+ExporterTest:
 	$(MAKE) -C ExporterTest
 
-ZAPDUtils/ZAPDUtils.a:
+.PHONY: ZAPDUtils
+ZAPDUtils:
```

Comment:
Adding those extra `.PHONY` isn't overriding the previous one?

Reply:
No, it seems to just treat it like a label for always being out-of-date rather than an actual target. I've seen other makefiles do this, and it seems to work here too. I thought it was worth keeping the submake ones separate from the main file's `.PHONY` targets.